### DISCUSSION TITLE
Point_stat outputs switched

### DISCUSSION THREAD

**Discussion by kadu-EPAGRI** [dtcenter/METplus#1904] [Statistical Computation] [answered]
Hi, I´m currently working with daily precipitation validation using ERA5 model and automatic surface observations from 2002 to 2022 at south Brazil. Its my first time using METplus and I noticed some swapped values in columns at point_stat outputs.  For example ,  FHO table output TOTAL = 1028534,  F_RATE = 0.9964,  H_RATE = 0.33754,  O_RATE =  0.33856.  It seems that the values ​​of F_RATE and H_RATE are switched.  The same happened with other point_stat tables. Any idea what could have happened? Has anyone ever had this kind of problem? Thanks in advance.

**Comment by kadu-EPAGRI** [comment]
It seems that the source of problem is the contingency table: FY_OY=347175;  FY_ON =677653; FN_OY=1045; FN_ON=2661. FY_ON and FN_ON are switched. Im running  METplus version V10.1.2.

**Reply by jprestop** [reply]
Hi @kadu-EPAGRI.  Thank you for your question.  I talked with a colleague for some guidance.  He told me that the FHO line type is equivalent to the CTC line type and suggested that you configure point_stat with both CTC and FHO output set turned on.  Then, please send us those two lines for the same case.  That will help us determine if there is an underlying bug.  

Also, in case it is helpful, here are some links to our FHO documentation:
[FHO H_RATE](https://met.readthedocs.io/en/latest/Users_Guide/appendixC.html?highlight=fho#h-rate)
[FHO Mean Forecast](https://met.readthedocs.io/en/latest/Users_Guide/appendixC.html?highlight=fho#mean-forecast)
[FHO Observation Forecast](https://met.readthedocs.io/en/latest/Users_Guide/appendixC.html?highlight=fho#observation-rate)

I hope this helps!

**Comment by kadu-EPAGRI** [comment]
Hi,
Thanks for your reply. Please find attached the  FHO and CTC stats files.
Regards
Carlos Eduardo Salles de Araujo, Dr.
Pesquisador EPAGRI / Ciram
+55 48 3665-5132


Em ter., 1 de nov. de 2022 às 15:25, jprestop ***@***.***>
escreveu:

> Hi @kadu-EPAGRI <https://github.com/kadu-EPAGRI>. Thank you for your
> question. I talked with a colleague for some guidance. He told me that the
> FHO line type is equivalent to the CTC line type and suggested that you
> configure point_stat with both CTC and FHO output set turned on. Then,
> please send us those two lines for the same case. That will help us
> determine if there is an underlying bug.
>
> Also, in case it is helpful, here are some links to our FHO documentation:
> FHO H_RATE
> <https://met.readthedocs.io/en/latest/Users_Guide/appendixC.html?highlight=fho#h-rate>
> FHO Mean Forecast
> <https://met.readthedocs.io/en/latest/Users_Guide/appendixC.html?highlight=fho#mean-forecast>
> FHO Observation Forecast
> <https://met.readthedocs.io/en/latest/Users_Guide/appendixC.html?highlight=fho#observation-rate>
>
> I hope this helps!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/dtcenter/METplus/discussions/1904#discussioncomment-4029915>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A35O26AOZ3DPFOFV26MF2ILWGFOBLANCNFSM6AAAAAARTJ7ISQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

VERSION MODEL     DESC FCST_LEAD FCST_VALID_BEG  FCST_VALID_END  OBS_LEAD OBS_VALID_BEG   OBS_VALID_END   FCST_VAR           FCST_UNITS FCST_LEV OBS_VAR OBS_UNITS OBS_LEV OBTYPE VX_MASK INTERP_MTHD INTERP_PNTS FCST_THRESH OBS_THRESH COV_THRESH ALPHA LINE_TYPE TOTAL   F_RATE H_RATE  O_RATE
V10.1.2 ERA5_LAND NA   000000    19900101_000000 19900101_000000 000000   20020101_000000 20220831_000000 Precipitation_Flux mm_d-1     *,*,*    PRATE   NA        A24     ADPSFC FULL    BEST_SQUARE 4           >=1.0       >=1.0      NA         NA    FHO       1028534 0.9964 0.33754 0.33856

VERSION MODEL     DESC FCST_LEAD FCST_VALID_BEG  FCST_VALID_END  OBS_LEAD OBS_VALID_BEG   OBS_VALID_END   FCST_VAR           FCST_UNITS FCST_LEV OBS_VAR OBS_UNITS OBS_LEV OBTYPE VX_MASK INTERP_MTHD INTERP_PNTS FCST_THRESH OBS_THRESH COV_THRESH ALPHA LINE_TYPE TOTAL   FY_OY  FY_ON  FN_OY FN_ON
V10.1.2 ERA5_LAND NA   000000    19900101_000000 19900101_000000 000000   20020101_000000 20220831_000000 Precipitation_Flux mm_d-1     *,*,*    PRATE   NA        A24     ADPSFC FULL    BEST_SQUARE 4           >=1.0       >=1.0      NA         NA    CTC       1028534 347175 677653  1045  2661

**Reply by jprestop** [reply]
Hi @kadu-EPAGRI.  

Thank you!  We can see the CTC counts followed by the FHO rates:
```
COL_NAME:   TOTAL  FY_OY  FY_ON FN_OY FN_ON EC_VALUE 
CTC: 1028534 347175 677653  1045  2661      0.5
COL_NAME:   TOTAL F_RATE  H_RATE  O_RATE  
FHO: 1028534 0.9964 0.33754 0.33856
```
These are fully derivable from each other, subject to roundoff error. And our team always recommend that people use CTC instead of FHO so as to avoid roundoff errors.

You can see the equations for how FHO and CTC are derivable from each other in the [parse_fho_ctable()](https://github.com/dtcenter/MET/blob/1f26c5524939021ff5d7c1606546e4b4625ebfde/src/tools/core/stat_analysis/parse_stat_line.cc#L51) function for Stat-Analysis. That function converts the FHO rates to the CTC counts and stores them that way.

Plugging in these counts/rates into those equations, we can see that:

- F_RATE = (FY_OY + FY_ON) / TOTAL = (347175 + 677653) / 1028534 = .996396813328 (which is rounded to 0.9964 in the FHO line)
- H_RATE = FY_OY / TOTAL = 347175 / 1028534 = .337543532834 (which is rounded to .33754 in the FHO line)
- O_RATE = (FY_OY + FN_OY) / TOTAL = (347175 + 1045) / 1028534 = .338559542028 (which is rounded to .33856)

So these numbers are definitely equivalent, and storing the contingency table counts as integers in the CTC line type clearly provides more accuracy than the rounded rates in the FHO line type.

I hope that helps.  Please let us know if you have additional questions.

**Comment by kadu-EPAGRI** [comment]
Thanks  for your reply.
The problem is in this result:
 F_RATE = (FY_OY + FY_ON) / TOTAL = (347175 + 677653) / 1028534 =
.996396813328 (which is rounded to 0.9964 in the FHO line)
Correct if i'm wrong but this says that the ERA5 (FY) estimates 99.6% of
rainy days in my region.  This seems to be very wrong.
 Also observe that H_RATE and O_RATE  values are very similar to each
other  (~0.33)  and better represent the real behavior of my region.
In my opinion the problem is in the value of FY_ON (677653) . To make some
realistic sense this value should be at least one hundred times smaller.
Which leads me to believe that 677653 is the value of FN_ON and not FY_ON.

Please, let me know if and where am i wrong.
Regards

Carlos Eduardo Salles de Araujo, Dr.

Pesquisador EPAGRI / Ciram
+55 48 3665-5132


Em ter., 1 de nov. de 2022 às 18:11, jprestop ***@***.***>
escreveu:

**Reply by JohnHalleyGotway** [reply]
Carlos,

I work with @jprestop in supporting the METplus tools. She and I talked about the questions you raised.

I see that you're getting some unexpected results related to the frequency of precip in your domain. These FHO, CTC, and corresponding CTS statistics are categorical. And by that I mean they are computed by first applying a threshold to the raw data to convert the precip amounts into yes/no (0/1) categories. So obviously, they are highly dependent on the choice of threshold.

In the sample data you sent, I see that you've chosen a threshold of `>=1.0`, and I see that you're processing 24-hour accumulations of precipitation. Of course, the units of the data being processed matter a lot, and its the user's job to make sure the units of the thresholds they choose are consistent with the units of the data being verified.

Typically, in GRIB1 and GRIB2 datasets, the units for precipitation are given in millimeters. If you're using GRIB inputs, any grid point with `>=1.0` millimeters of precipitation is counted as an event. So the first step, is to double-check your interpretation based on the units of the data you're verifying.

Next, I think it's always a great idea to use MET's `plot_data_plane` tool to visualize some sample data for a single run. I obviously don't have the data you're using, but I'll demonstrate with some sample data included in the MET releases:
```
bin/plot_data_plane data/sample_fcst/2005080700/wrfprs_ruc13_24.tm00_G212 APCP_24.ps 'name="APCP"; level="A24";' -plot_range 0 1
```
Note that I used `-plot_range 0 1` so that the colorbar tops off at 1.0 (which is the threshold you're using). So anything appearing pink in the following plot is >= 1.0 and would count as an event:
<img width="591" alt="Screen Shot 2022-11-03 at 11 41 14 AM" src="https://user-images.githubusercontent.com/21087144/199795419-7dbac676-7b1b-41f1-a911-fdb845ad26d5.png">

You see that a large portion of the region of valid data does meet that threshold criteria. So I'd recommend:
1. Double-checking the units for precip in your dataset.
2. Run `plot_data_plane` on some sample cases to see what portion of the grid points meet the `>=1.0` threshold.
3. Try running with additional thresholds by setting `cat_thresh = [ >=1.0, >=6.35, >=12.7, >=25.4 ];` Then check to see how the rates and statistics vary as you increase the precip threshold.

Since 25.4 mm = 1.0 inch, in the US we often use these thresholds corresponding to 1/4 inch, 1/2 inch, and 1.0 inches.

Hope that helps clarify.

**Comment by kadu-EPAGRI** [comment, marked as answer]
Hi John
Finally I found the problem with point_stats results.  At the beginning of
the process I used CDO to merge all ERA5 timestep files in one big .nc
file. I also used MET  ASCII2NC to create a unique observational .nc file
with all timesteps . When I ran point_stats the routine wrongly matched
just a fixed timestep of ERA5 with all observations timesteps. I fixed it
using  point_stats with individual  ERA5 timestep files to compute matched
pairs and  STAT_ANALYSIS to compute all the statistics.
A final question: Point_stats is able to match date/time in the way I first
tried, using 2 big .nc files (observation and forecast)?  Maybe using the
reftime parameter/dimension  in the ERA5 big .nc file ( originally it only
had the time dimension) ?
Regards

Carlos Eduardo Salles de Araujo, Dr.
Pesquisador EPAGRI / Ciram
+55 48 3665-5132

**Reply by JohnHalleyGotway** [reply]
Carlos,

I'm glad to hear you were able to make progress and uncover the issue. While I understand the gist of what you describe, I don't fully appreciate the details. But let me describe how time filtering is handled in Point-Stat.

You run Point-Stat by passing to it a gridded forecast file, one or more files containing point observations, and a configuration file. The configuration file defines what data should be extracted from the forecast file and how it should be compared to the point observations.

For each field listed in the configuration file, Point-Stat extract the gridded forecast data and check the VALID TIME of that data. By default, it defines a matching observation time window relative to that VALID TIME using the [obs_window](https://github.com/dtcenter/MET/blob/61e192883f76bf7362152c19c2d279fa9d5dc5c7/met/data/config/PointStatConfig_default#L195) config entry. It's described in the user's guide in [this section](https://met.readthedocs.io/en/latest/Users_Guide/config_options.html?highlight=obs_window#obs-window).

By way of example, if you forecast is valid at 20220807_120000 and set:
```
obs_window = { beg = -10800; end = 10800; }
```
Point-Stat will use point observations with timestamps between 20220807_09 and 20220807_15 because 10,800 seconds = 3 hours.

However, if you set the `-obs_valid_beg` and `-obs_valid_end` command line options to define the matching time window, then those settings override the config file settings.

So yes, you certainly should be able to filter obs in Point-Stat based on their valid time. And I'd recommend running at logging verbosity level 3 (e.g. `-v 3`). That'll print out log messages for each verification task that look like this:
```
DEBUG 3: Number of matched pairs   = 3967
DEBUG 3: Observations processed    = 89893
DEBUG 3: Rejected: station id      = 0
DEBUG 3: Rejected: obs var name    = 82886
DEBUG 3: Rejected: valid time      = 0
DEBUG 3: Rejected: bad obs value   = 0
DEBUG 3: Rejected: off the grid    = 4
DEBUG 3: Rejected: topography      = 0
DEBUG 3: Rejected: level mismatch  = 2902
DEBUG 3: Rejected: quality marker  = 0
DEBUG 3: Rejected: message type    = 134
DEBUG 3: Rejected: masking region  = 0
DEBUG 3: Rejected: bad fcst value  = 0
DEBUG 3: Rejected: bad climo mean  = 0
DEBUG 3: Rejected: bad climo stdev = 0
DEBUG 3: Rejected: mpr filter      = 0
DEBUG 3: Rejected: duplicates      = 0
```
And you can read this [Q&A answer](https://met.readthedocs.io/en/latest/Users_Guide/appendixA.html#q-how-can-i-understand-the-number-of-matched-pairs) for an explanation of how to interpret all those counts.

Hope that helps.

**Reply by JohnHalleyGotway** [reply]
Just wanted to follow up on this discussion to see if questions remain. If not, please feel free to select the "Mark As Answer" button for one of the previous responses, and then I'll lock it to prevent future posts. If I don't hear otherwise in the next week or so, I'll just go ahead and mark it as being answered.

Thanks!
